lnrpc/invoicesrpc/addinvoice.go

```diff
@@ -678,169 +619,233 @@ type SelectHopHintsCfg struct {
 }
 
 // sufficientHints checks whether we have sufficient hop hints, based on the
-// following criteria:
-//   - Hop hint count: limit to a set number of hop hints, regardless of whether
-//     we've reached our invoice amount or not.
-//   - Total incoming capacity: limit to our invoice amount * scaling factor to
-//     allow for some of our links going offline.
+// any of the following criteria:
+//   - Hop hint count: the number of hints have reach our max target.
+//   - Total incoming capacity: the sum of the remote balance amount in the
+//     hints is bigger of equal than our target (currently twice the invoice
+//     amount)
 //
 // We limit our number of hop hints like this to keep our invoice size down,
 // and to avoid leaking all our private channels when we don't need to.
-func sufficientHints(numHints, maxHints, scalingFactor int, amount,
-	totalHintAmount lnwire.MilliSatoshi) bool {
+func sufficientHints(nHintsLeft int, currentAmount,
+	targetAmount lnwire.MilliSatoshi) bool {
 
-	if numHints >= maxHints {
-		log.Debug("Reached maximum number of hop hints")
+	if nHintsLeft <= 0 {
+		log.Debugf("Reached targeted number of hop hints")
 		return true
 	}
 
-	requiredAmount := amount * lnwire.MilliSatoshi(scalingFactor)
-	if totalHintAmount >= requiredAmount {
+	if currentAmount >= targetAmount {
 		log.Debugf("Total hint amount: %v has reached target hint "+
-			"bandwidth: %v (invoice amount: %v * factor: %v)",
-			totalHintAmount, requiredAmount, amount,
-			scalingFactor)
-
+			"bandwidth: %v", currentAmount, targetAmount)
 		return true
 	}
 
 	return false
 }
 
-// SelectHopHints will select up to numMaxHophints from the set of passed open
-// channels. The set of hop hints will be returned as a slice of functional
-// options that'll append the route hint to the set of all route hints.
-//
-// TODO(roasbeef): do proper sub-set sum max hints usually << numChans.
-func SelectHopHints(amtMSat lnwire.MilliSatoshi, cfg *SelectHopHintsCfg,
-	openChannels []*HopHintInfo,
-	numMaxHophints int) [][]zpay32.HopHint {
-
-	// We'll add our hop hints in two passes, first we'll add all channels
-	// that are eligible to be hop hints, and also have a local balance
-	// above the payment amount.
-	var totalHintBandwidth lnwire.MilliSatoshi
-	hopHintChans := make(map[wire.OutPoint]struct{})
-	hopHints := make([][]zpay32.HopHint, 0, numMaxHophints)
-	for _, channel := range openChannels {
-		enoughHopHints := sufficientHints(
-			len(hopHints), numMaxHophints, hopHintFactor, amtMSat,
-			totalHintBandwidth,
-		)
-		if enoughHopHints {
-			log.Debugf("First pass of hop selection has " +
-				"sufficient hints")
+// populateForcedHints check that the provided hint paths are valid and returns
+// an array of hop hints.
+func populateForcedHints(forcedHints [][]zpay32.HopHint) ([][]zpay32.HopHint,
+	error) {
 
-			return hopHints
+	hopHints := make([][]zpay32.HopHint, 0, maxHopHints)
+	for _, hint := range forcedHints {
+		if len(hint) == 0 {
```

Comment:
I wonder if this check can't just be enforced at a higher level (potentially during RPC parsing)? Why would a forced hint (provided by the user) ever have zero length?

Reply:
we can check at a higher level :+1: